ecosystem/sep-0010.md

```diff
@@ -107,8 +107,8 @@ To validate the challenge transaction the following steps are performed by the s
 * verify that transaction source account is equal to the server's signing key;
 * verify that transaction has time bounds set, and that current time is between the minimum and maximum bounds;
 * verify that transaction contains a single [Manage Data](https://www.stellar.org/developers/guides/concepts/list-of-operations.html#manage-data) operation and it's source account is not null;
-* verify that transaction envelope has a correct signature by server's signing key;
-* verify that transaction envelope has a correct signature by the operation's source account;
+* verify that transaction envelope has signatures from valid signers of the transaction's source account with summed weight greater than the medium threshold on the transaction's source account;
+* verify that transaction envelope has signatures from valid signers of the operation's source account with summed weight greater than the medium threshold on the operation's source account;
```

Comment:
What is the reasoning behind threshold level selection here? I believe that depending on the use case it might be low, medium or high, which leads me to think that it might be better to move such check on a higher level. Authentication process goal is to verify that certain request comes from a user who has a private key of a certain public key, and threshold check looks more like authorization (verifying that the identified user has privileges to perform certain action on the account), which is outside of the scope for this SEP.

Reply:
Well, the issue is that this SEP is not really possible to implement (from my POV) without some sort of specification on which sigs are needed by the client. I do agree that if we are being complete, we should specify exactly which threshold is needed for each operation. However, I think it was mostly for time's sake the medium threshold was chosen.

Reply:
In other words, there are many users that have multisig accounts of various forms. How does this SEP deal with them? Do we just produce a token (and allow a user to change KYC docs) for a 1/3 multisig signature? That doesn't seem right...

Reply:
The way I see it is that this SEP operates primarily with keypairs, and in general doesn't require an associated Stellar account to exist. For example, it can be used in a centralized multi-sig coordinator (simple web service where you authenticate with a particular keypair and receive a  list of transactions pending signature by your keypair), or there might be a case where corresponding account doesn't exist yet (like the service allows withdrawals to non-funded account via `create_account` operation, but wants to ensure that funds wont be locked out because of a typo, or stollen if some malware tampered with destination address).

I think one possible improvement is to support auth with several keypairs at once (like you pass a comma-separated list of public keys when requesting the challenge and then on the auth stage service can check that the tx is signed by all those keys). But since this SEP is based on `wallet <-> web-service` interaction over relatively short period of time, I don't quite see a real use-case for multi-sig on the auth transaction, because the wallet would have to collect all required signatures a) very quickly (recommended expiration for the challenge tx is 5 minutes); and b) co-signers would have to sign very generic transaction "allow A to perform operations with our shared account X on service S", with no granularity over operations and basically no way to revoke such permission until auth session expires.

> However, I think it was mostly for time's sake the medium threshold was chosen.

My point was that since this SEP defines a generic authentication (not _authorization_) protocol, it doesn't seem to be a proper place to define specific thresholds at all. Authorization logic is generally app-specific but can be defined in higher-level SEPs when necessary.

Reply:
I've updated the text to include the case of the account not existing on pubnet (in which case, the client just needs to send the signature by the master key).

To your point, I understand that this is just an authentication SEP. However, the problem is that right now it is specified as the authentication standard for SEP-6 and SEP-12. Currently, the way this SEP is written, the overall system is broken (because clients are getting authenticated for accounts where they shouldn't be).

I agree with you that an authorization SEP is the right way to do this (where you have application specific tokens and different permission levels etc.) I don't have the time to solve this elegantly in a separate SEP (thus the medium threshold specification). If you do, please go ahead.

I am just a SEP-6 and SEP-12 implementer and simply cannot implement as the whole thing currently exists.

Reply:
@nebolsin @istrau2 This is a great discussion.

@nebolsin you make a great point about this protocol needing to function with key pairs that don't have associated Stellar accounts. I think we should be sure to preserve that and state it as a goal of the protocol in the Abstract section.

I think the use case of multi-sig accounts is important, and will become more common. I'm thinking of users of services like StellarGuard or users with SEP-8 accounts that include one or more issuers as signers. I think the best approach is for this protocol to specify a sensible default way to handle created Stellar accounts (using medium threshold), and make it clear that different applications can choose different thresholds or ways of handling multi-sig accounts. That maximizes the simplicity of building on SEP-10, while also allowing flexibility in certain cases where medium threshold isn't appropriate.

I'm not too worried about the recommended time it takes to sign a transaction appropriately. Implementers can bump it up if it's causing an issue in their case, and automatic signing like SEP-8 will be more and more common.

Thoughts?

Reply:
@tomquisel I totally agree that multi-sig use case is very important, I'm just not convinced that it should be included into this SEP as a requirement. And as an optional feature it's already allowed by the last item in verification rules so there's no conflict between this SEP and multi-sig: 

> * use operations's source account to determine the authenticating client and perform *any additional service-specific validations*

We can definitely add a note for implementors describing the security considerations for multi-sig case and some hints about recommended setup. 

But let me stop right here for now, because I just realized that even with proposed changes this SEP will not work for StellarGuard case, because as far as I know it receives partially-signed tx, adds a final missing signature after 2FA check and submits the transaction to the network. And this SEP intentionally makes challenge transaction invalid so it will fail if submitted. And for now I don't see a clear solution for this scenario, need to think more about it. Any ideas?

Reply:
@tomquisel Another consideration against mixing authentication and authorization here is that JWT token obtained via this SEP is a session token and it will remain valid for some prolonged period (currently the spec advises to use 24 hours). But during that time account multi-sig configuration on Stellar could change, for example a person who requested a JWT token could be excluded from co-signers or his weight is adjusted. If we perform authorization as part of authentication, the JWT will encode account multi-sig configuration at issue time and it will remain static throughout token lifetime, which have pretty bad security implications.

I think that's a pretty strong argument to keep authentication focused on identity (so that JWT token only claims ability to sign transactions on behalf of `sub` keypair), while authorization, including multisig checks, should be enforced on a service level for every particular action. 

For example, in case of SEP-12 `/customer` endpoint handler should first extract authenticated keypair from JWT and verify that it is authorized to perform the requested action on the account. To support multi-sig setups it should either accept multiple JWT tokens, or we can extend this SEP to allow single JWT to claim control of multiple keypairs. I'll try to draft a PR for this latter approach.

Reply:
@nebolsin
I have to say that I disagree with your last post, authorization should be included in the JWT, whether or not it is included in the SEP is a preference thing. Including basic authorization related claims (such as groups, roles and permissions) is a standard, widely adopted practice. This is because centralizing authorization logic (at least the generic, non-application specific kind of logic) is a really good idea, so that one can have a focused authentication/authorization service that is responsible for this kind of thing. For example, think of what google, facebook etc. do when you request an OAuth2 JWT from them, they prompt the user to select which permissions to allow to the app.

More to the point, for our (i.e. Stellarport's) specific implementation of this SEP, we need to do authentication and authorization in one place. Our KYC server is not capable of doing the basic authorization (i.e. determining which thresholds are met by certain signatures) because it is not connected to a Stellar node.

In terms of some sort of permissions changing on the underlying account, this is a general problem with JWTs, not really specific to our use case. Any time a service issues a JWT, the underlying permissions on the account might change. That change will not be reflected in the JWT. The way to solve this is by limiting the JWT to a short enough period of validity. In fact, I am not sure why this spec advises a validity period, I would think that is very application specific.

Perhaps the solution here is for the JWT to include an optional roles claim which can include Master, Low, Medium and High roles. Master means that the token guarantees that this user has signed with the master key. The other three correspond to the various Stellar account thresholds. In the case that there is no roles claim, the consuming application should assume the Master claim, this makes the change in the SEP backwards compatible.

What do you think?

Reply:
I have modified the PR with the optional roles claim instead.